text/0000-app-image-extensions.md

```diff
@@ -0,0 +1,243 @@
+
+The advantages of using a buildpack for privileged operations are the same as for unprivileged operations: they are composable, fast (caching), modular, reuseable, and safe.
+
+As an example, consider the use case of Acme.com. They have three teams that want to contibute privileged layers to every image built at the company. Each of these teams wants to manage their own artifacts and the mechanisms for installing them, but none of these teams are in control of the base image. The teams have special logic for when their layers should be added to an image; in some cases it's the precense of a Node.js app, and in others it's the use of Tomcat, etc. In the past, these teams were trying to contribute their layers by selectively adding them to every `Dockerfile` for every base image in the company, but this doesn't scale well. With Root Buildpacks, each team can manage their own artifacts, release cadence, and add logic to `bin/detect` to ensure the layers are added when needed.
+
+# What it is
+[what-it-is]: #what-it-is
+
+- *root buildpack* - a special case of buildpack that is run with privileges
+
+Application developers can use root buildpacks to extend their build and/or run images.
+
+We introduce a boolean `privileged` key in the `[buildpack]` table of `buildpack.toml`, which is defined as follows:
+
+```
+[buildpack]
+privileged = <boolean (default=false)>
+```
+
+When `privileged` is set to `true`, the lifecycle will run this buildpack as the `root` user.
+
+For each root buildpack, the lifecycle will use [snapshotting](https://github.com/GoogleContainerTools/kaniko/blob/master/docs/designdoc.md#snapshotting-snapshotting) to capture changes made during the buildpack's build phase (excluding `/tmp`, `/cnb`, and `/layers`). Alternatively to snapshotting, a platform may store a new stack iamge to cache the changes. All of the captured changes will be included in a single layer produced as output from the buildpack. The `/layers` dir MAY NOT be used to create arbitrary layers.
+
+The buildpack can then exclude directories from the layer by writing to the `launch.toml` using a new `[[excludes]]` table, for example:
+
+```toml
+[[excludes]]
+paths = [ "/var" ]
+cache = true
+```
+
+* `paths` - a list of path globs to exclude from the run image
+* `cache` - if set to `true`, the paths will be stored in the cache
+
+The `[process]` and `[[slices]]` tables can be used as normal.
+
+The following constraint(s) will be enforced:
+* If a user attempts to create a buildpackage including both root buildpacks and non-root buildpacks, the process will fail.
```

Comment:
What's the rationale behind this constraint? What is the concern it's addressing?

Reply:
This addresses a concern that @sclevine had about root buildpacks proliferating through the ecosystem. Like, if we give this tool to people, they will use it too frequently. If we instead restrict its use, it forces buildpacks authors to explore alternative solutions.

I don't necessarily agree with that concern, but I'm willing to accept it in compromise (in part because it's easy to reverse).